dbms/src/Flash/executeQuery.cpp

```diff
@@ -12,27 +12,96 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <Common/FailPoint.h>
+#include <Common/ProfileEvents.h>
+#include <Flash/Coprocessor/DAGContext.h>
 #include <Flash/Coprocessor/DAGQuerySource.h>
 #include <Flash/Planner/PlanQuerySource.h>
 #include <Flash/executeQuery.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/ProcessList.h>
+#include <Interpreters/Quota.h>
 #include <Interpreters/executeQuery.h>
 
+namespace ProfileEvents
+{
+extern const Event Query;
+}
+
 namespace DB
 {
-BlockIO executeQuery(
-    Context & context,
-    bool internal,
-    QueryProcessingStage::Enum stage)
+namespace FailPoints
+{
+extern const char random_interpreter_failpoint[];
+} // namespace FailPoints
+namespace
+{
+void prepareForExecute(Context & context)
+{
+    ProfileEvents::increment(ProfileEvents::Query);
+    context.setQueryContext(context);
+
+    QuotaForIntervals & quota = context.getQuota();
+    quota.addQuery(); /// NOTE Seems that when new time interval has come, first query is not accounted in number of queries.
+    quota.checkExceeded(time(nullptr));
+}
+
+BlockIO executeDAG(IQuerySource & dag, Context & context, bool internal)
+{
+    RUNTIME_ASSERT(context.getDAGContext());
+    auto & dag_context = *context.getDAGContext();
+    const auto & logger = dag_context.log;
+    RUNTIME_ASSERT(logger);
+
+    prepareForExecute(context);
+
+    if (!internal)
+        logQuery(dag.str(context.getSettingsRef().log_queries_cut_to_length), context, logger);
+    FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::random_interpreter_failpoint);
+    auto interpreter = dag.interpreter(context, QueryProcessingStage::Complete);
+    BlockIO res = interpreter->execute();
+
+    // For internal == true, no need to set up the resource usage tracker.
+    if (!internal)
+    {
+        /// Hold element of process list till end of query execution.
+        if (dag_context.is_mpp_task)
+        {
+            /// for MPPTask, process list entry is created in MPPTask::prepare()
+            RUNTIME_ASSERT(dag_context.getProcessListEntry() != nullptr, "process list entry for MPP task must not be nullptr");
+            res.process_list_entry = dag_context.getProcessListEntry();
+        }
+        else
+        {
+            RUNTIME_ASSERT(dag_context.getProcessListEntry() == nullptr, "process list entry for non-MPP must be nullptr");
+            res.process_list_entry = setProcessListElement(
```

Comment:
Why `setProcessListElement` after `interpreter->execute()`

Reply:
I think order doesn't matter here, just want to reduce one variable process_list_entry
But I can keep the same order as the original implementation.

Reply:
o.. may be relevant, such as starting an asynchronous read task in `interpreter->execute()`...

Reply:
fixed.